Speculative query execution (JAVA-561)

olim7t opened the pull request:
Preemptively query the next host when the current one takes too long to respond. Introduces a new `SpeculativeExecutionPolicy` (I deliberately avoided the word "retry" which was already used in driver jargon).

Most of the `RequestHandler` logic was extracted in a `SpeculativeExecution` inner class. A RequestHandler starts multiple SpeculativeExecutions which compete for the same query plan. This allows:
- coherent integration with the retry logic: each SpeculativeExecution can still retry, as defined by the `RetryPolicy`.
- few changes to existing synchronization code. Each SpeculativeExecution manages its own requests with the existing `QueryState` mechanism, and the RequestHandler coordinates all SpeculativeExecutions.

olim7t committed:
Speculative query executions (JAVA-561).

olim7t commented:
This is an early preview. Still on my todo list:
- expose number of speculative executions as a metric
- maybe add the ability to set custom key/value pairs on `Statement`, which could be used for custom logic in `SpeculativeExecutionPolicy` (this would be useful for all policies)

pcmanus commented:
> maybe add the ability to set custom key/value pairs on `Statement`

Fyi, my idea (which I never took the time to add unfortunately) was to let people implement their own `Statement` if they want to ship custom information.  Now, this is easy for `RegularStatement` but not really for the other ones, so I think we could have an abstract `ForwardingStatement` class that wraps any other statement and allow to customize it. Imo that's a little bit cleaner and generic that custom key/value pairs (I also feel that it separate concerns better API wise).

olim7t commented:
Yes, I agree that wrapping is more flexible. I added a commit for that.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                                retry(false, null);
+                                return;
+                            case UNPREPARED:
+                                // Do not release connection yet, because we might reuse it to send the PREPARE message (see write() call below)
```
So this is the main reason why we released later for [JAVA-633](https://datastax-oss.atlassian.net/browse/JAVA-633) right?  So the net of the change for JAVA-666 is to release in all cases of onSet except when you get an UNPREPARE error and the prepared statement will be reprepared.  Just making sure I'm understanding correctly.

olim7t replied:
Yes, that's it.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                return false;
+
+            if (nextExecutionScheduled.compareAndSet(false, true))
+                scheduleExecution(speculativeExecutionPlan.nextExecution(host));
```
Do we plan on adding some kind of control to SpeculativeExecutionPolicy like we do for RetryPolicy with RetryDecision?  My first expectation was that speculative retries would only be for reads, but I can understand why you would want to do it for mutations to in some cases.

I am thinking for the same reason we have onWrite/onReadTimeout in RetryPolicy that we'll want to control whether or not we schedule another execution.  Some scenarios I can think of:
1. counter updates (if a speculative retry gets executed and the first execution does eventually succeed, you'll increment a counter twice).
2. mutations in general.  i.e. if an update gets speculative retried and succeeds, user gets the confirmation and performs another update (for some reason), and then the first update from the first query succeeds (maybe it was slowed down by network latency on a connection).  If the user understands this could happen it may not be a big deal.

olim7t replied:
The problem with speculative retries is that they happen before we've received a response from the server, so we might not know anything about the statement (for example if it's a SimpleStatement, since we don't parse query strings client-side).
That's the reason why I introduced `StatementWrapper`. Users who want more precise control would need to "mark" the relevant statements and have special cases in a custom policy.
I don't see a way to provide that in a generic way in the driver.

tolbertam commented:
> coherent integration with the retry logic: each SpeculativeExecution can still retry, as defined by the RetryPolicy.

At first glance it is slightly confusing to me that there are effectively 2 separate policies for re-executing a query.  My interpretation of a SpeculativeExecution is that it is a form of Retry, and that the previous query execution should not be retried using a RetryPolicy, but I understand that a RetryPolicy is in regards to reacting to cassandra timeouts, where a speculative execution is for reacting to no response being received within some client timeout.  I may understand this better when I review the code more closely.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                                    retry(true, retry.getRetryConsistencyLevel());
+                                    break;
+                                case RETHROW:
+                                    setFinalResult(connection, response);
```
One issue here (although maybe not a problem) is that if you have subsequent speculative retries on this RequestHandler and the first request or one of the requests rethrows won't that mark the request as done while other attempts are in progress?  That could eliminate some of the utility the other speculative retries.

olim7t replied:
As far as RequestHandler is concerned, the goal is to get a definitive answer to the query. Whether that answer is a result or an exception makes no difference. The first execution that gives us this answer wins, and the others will be cancelled (the underlying assumption being that if an execution fails, the others would probably fail too).

tolbertam replied:
That sounds good to me, thanks for clearing that up.  So a user could actually still see a ReadTimeoutException if speculative executions are in progress, but I can see why you wouldn't want to wait for the other queries to complete as you could see the same thing later.   The speculative execution time I imagine is going to typically be very low (milliseconds) which is should be less than the cassandra read timeout by some large factor.

olim7t commented:
> At first glance it is slightly confusing to me that there are effectively 2 separate policies for re-executing a query. My interpretation of a SpeculativeExecution is that it is a form of Retry, and that the previous query execution should not be retried using a RetryPolicy

My motivation for this was the following scenario:
- you 're using `DowngradingConsistencyRetryPolicy`. You're doing a query at CL = TWO but the first host gets a read timeout after 200 ms, that says that only one replica responded.
-  you've set the speculative execution limit at 150 ms.

If speculative executions cancel regular retries, the first execution will never get the chance to do the retry at CL=ONE. So you're throwing away a retry that might have succeeded, in favor of a speculative execution that might get the same read timeout and keep failing.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -16,168 +16,125 @@
     private final Iterator<Host> queryPlan;
+    private final SpeculativeExecutionPlan speculativeExecutionPlan;
+    private final Set<SpeculativeExecution> runningExecutions = Sets.newHashSet();
+    private final Set<Timeout> scheduledExecutions = Sets.newHashSet();
```
Shouldn't these sets be concurrent / externally synchronized? In particular, there are iterations over them in `cancelPendingExecutions()` which, as long as I could see, could be executed while other threads are modifying the sets.

olim7t replied:
Yes, I've fixed it.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -16,168 +16,125 @@
+        if (isDone.get() || delayMillis <= 0)
+            return;
+        logger.trace("[{}] Schedule next speculative execution in {} ms", id, delayMillis);
+        scheduledExecutions.add(scheduler.newTimeout(newExecutionTask, delayMillis, TimeUnit.MILLISECONDS));
```
Isn't there a slight risk here that the new task gets executed _before_ it is added to the set?

olim7t replied:
That could happen with an extremely low speculative delay, which is unlikely. But anyway, leaving a completed `Timeout` in `scheduledExecutions` is not really a problem, the worst that can happen is that we cancel it when it's done, which is a no-op.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/StatementWrapper.java:
```diff
@@ -0,0 +1,152 @@
+ * }
+ * </pre>
+ */
+public abstract class StatementWrapper extends Statement {
```
This new class will cause a problem with QueryLogger. QueryLogger only handles the 3 pre-defined types of Statement: Regular, Bound and Batch. You should at least add an if-block for StatementWrapper instances in `QueryLogger#append(Statement, StringBuilder, int)` and "unwrap" the statement to log. 
As a side note, IMHO the method `getQueryString()`, which is currently defined twice in PreparedStatement and RegularStatement, should be pulled up into Statement class and be part of its API. Implementations would be trivial for most subclasses, except maybe BuiltStatement.

olim7t replied:
Yes, I need to change `QueryLogger` to do the same as in `SessionManager#makeRequestMessage`, thanks for pointing that out.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/SpeculativeExecutionTest.java:
```diff
@@ -0,0 +1,218 @@
+        ccm.waitForDown(1);
+
+        // Put an SCassandra instance in host1's place
+        String address = CCMBridge.ipOfNode(1);
```
The following _might_ spare you the use of Scassandra: 

In CCMBridge, add:

``` java

    public void pause(int n) {
        logger.info("Pausing: " + IP_PREFIX + n);
        execute("ccm node%d pause", n);
    }

    public void resume(int n) {
        logger.info("Resuming: " + IP_PREFIX + n);
        execute("ccm node%d resume", n);
    }
```

Then in your test:

``` java
    @Test(groups = "short")
    public void should_start_speculative_execution_if_first_execution_takes_too_long() throws InterruptedException {
        loadBalancingPolicy.order = ImmutableList.of(1, 2, 3);
        long startCount = speculativeExecutions.getCount();
        ResultSet rs;
        try {
            ccm.pause(1);
            rs = session.execute("select v from foo where k = 1");
            Thread.sleep(2000);
        } finally {
            ccm.resume(1);
        }
        Row row = rs.one();
        assertThat(row.getString("v")).isEqualTo("value from real host");
        assertThat(speculativeExecutions.getCount()).isEqualTo(startCount + 1);
        assertThat(rs.getExecutionInfo().getQueriedHost()).isEqualTo(TestUtils.findHost(cluster, 2));
    }
```

At least, it works on my laptop :)

olim7t replied:
Nice, I hadn't noticed that these commands had been added to CCM. They will make things simpler indeed.

olim7t commented:
Added idempotency on statements. There's work left to do on QueryBuilder so that BuiltStatements set their flag automatically (based on whether they update counters or prepend/append to lists).
The API is a bit complicated, but I wanted the ability for a user to say: "I know my app never updates counters or prepends to lists, so I want all statements to be idempotent by default" (right now the out-of-the-box default is non-idempotent). I hope my javadocs make sense.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/policies/SpeculativeExecutionPolicy.java:
```diff
@@ -0,0 +1,69 @@
+import com.datastax.driver.core.Statement;
+
+/**
+ * The policy that decides if the driver will send speculative queries to the next hosts when the current host takes too
```
A minor comment: speculative execution does not take place "when" the host takes too long to respond. IMHO this should be something along the lines of : "...if the driver should send the same query to more than one host, to optimize response delays and defend itself against hosts that take too long to respond."
You could also point users to http://en.wikipedia.org/wiki/Speculative_execution.

olim7t replied:
Sorry but I don't get it :)
Speculative executions are triggered when the current host has not replied after x milliseconds, that's what I mean by "too long".

Looking at this comment I realize that it should mention idempotency, I'm adding it.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/Statement.java:
```diff
@@ -326,4 +326,49 @@ public Statement setPagingState(PagingState pagingState) {
+     * @return whether this statement is idempotent, or {@code null} to use
+     * {@link QueryOptions#getDefaultIdempotency()}.
+     */
+    public Boolean isIdempotent() {
```
I'm slightly disturbed here by the asymmetric getter/setter pair. If idempotency is a property of statements, a statement is either idempotent or non-idempotent, so null values sound strange and conceptually would mean "I don't know if I'm idempotent" or "I may be idempotent". 
I understand that this comes from the need to introduce a "default" idempotency cluster-wide. 
But in this case the message we are sending here to the receiver slid from the initial intent: "Are you idempotent?" to: "Do you allow speculative execution?" - in which case it would be fine to return null (meaning "I don't know if I can be executed speculatively, ask the cluster").
I would suggest to make the API clearer:
- Renaming the property to `speculativeExecutionPermission` or stg in the style
- Instead of Boolean, use an enum like `SpeculativeExecutionPermission { DEFAULT, ALLOW, DENY }`

newkek replied:
I agree that this method should more express the speculative aspect of the statement than its idempotency even though these words have the same meaning in the driver, if this is what you mean Alexandre. Idempotency is more an internal technical aspect in the driver and I think speculativeExecutionPermission is clearer concerning the impact on the behaviour of the driver for the clients.

olim7t replied:
> null values sound strange and conceptually would mean "I don't know if I'm idempotent"

That's pretty much what it is. The driver does not get in the business of parsing queries, so given a `SimpleStatement` it doesn't know if it's idempotent or not.

> Idempotency is more an internal technical aspect in the driver

Actually I disagree. Idempotency is a intrinsic concept and I think it's quite understandable. And we might end up using in other places than speculative execution.

tolbertam commented:
Updated licenses to 2015 for a few files that didn't have it to get the maven build to pass, feel free to squash into commits when you go to do it.

tolbertam commented:
Noticed a noisy log entry at debug, reduced to trace and also check for trace before logging (not sure if needed but seemed like a good idea).

tolbertam committed:
Add test for JAVA-666 that validates stream id release timing.

ConnectionReleaseTest validates that when a future is set and its
listeners are called that the connection associated with that future is
released beforehand, so if a listener tied to that future calls executeAsync
it is more likely able to acquire a connection immediately.

olim7t commented:
Andy, I've slightly changed your fix to avoid sharing the `Request` object: instead of doing it each time in `request()`, I do it before creating each execution, and it takes it as a constructor parameter.

I've also addressed the thread-safety issues pointed out by Alexandre. Because iterators aren't safe by consumption by multiple threads (even when they come from concurrent collections), I've added a synchronized wrapper around the query plan. My tests show zero contention on the monitor, so I think synchronized is good enough.

tolbertam commented:
:+1: on those changes @olim7t,  I like that the request is now generated and passed into the execution, definitely seems more safe to future changes (i.e. can now handle request() being called multiple times for the same execution).

olim7t commented:
Rebased on 2.0 + squashed commits.
